File: src/linear_algebra_diagonalization_easy.py
"""
linear_algebra - diagonalization (easy)
Generated: 2026-02-22T05:59:32.910958
"""

from problem_utils import *

def generate():
    prob_type = randint(1, 4)
    
    if prob_type == 1:
        # Type 1: Check if a matrix is diagonal (1000-1100)
        # Generate a diagonal matrix and ask if it's diagonal
        n = randint(2, 3)
        diag_vals = [nonzero(-5, 5) for _ in range(n)]
        M = Matrix.diag(*diag_vals)
        
        return problem(
            question=f"Is the matrix $A = {latex(M)}$ a diagonal matrix? (Answer true or false)",
            answer=True,
            difficulty=(1000, 1100),
            topic="linear_algebra/diagonalization",
            solution=steps(
                "A matrix is diagonal if all off-diagonal entries are zero",
                f"Matrix $A$ has entries ${latex(M)}$",
                "All off-diagonal entries are 0",
                "Therefore, $A$ is diagonal"
            ),
        )
    
    elif prob_type == 2:
        # Type 2: Find eigenvalues of a 2x2 diagonal matrix (1100-1200)
        # For diagonal matrices, eigenvalues are the diagonal entries
        d1, d2 = nonzero(-5, 5), nonzero(-5, 5)
        M = Matrix([[d1, 0], [0, d2]])
        eigenvals = sorted([d1, d2])
        ans = FiniteSet(*eigenvals)
        
        return problem(
            question=f"Find all eigenvalues of the diagonal matrix $A = {latex(M)}$",
            answer=ans,
            difficulty=(1100, 1200),
            topic="linear_algebra/diagonalization",
            solution=steps(
                "For a diagonal matrix, the eigenvalues are the diagonal entries",
                f"The diagonal entries are ${d1}$ and ${d2}$",
                f"Therefore, the eigenvalues are ${latex(ans)}$"
            ),
        )
    
    elif prob_type == 3:
        # Type 3: Check if a given vector is an eigenvector of a diagonal matrix (1200-1300)
        d1, d2 = nonzero(-5, 5), nonzero(-5, 5)
        while d1 == d2:
            d2 = nonzero(-5, 5)
        
        M = Matrix([[d1, 0], [0, d2]])
        
        # Create a standard basis vector as eigenvector
        if randint(0, 1) == 0:
            v = Matrix([1, 0])
            eigenval = d1
        else:
            v = Matrix([0, 1])
            eigenval = d2
        
        Av = M * v
        
        return problem(
            question=f"Is $v = {latex(v)}$ an eigenvector of $A = {latex(M)}$? (Answer true or false)",
            answer=True,
            difficulty=(1200, 1300),
            topic="linear_algebra/diagonalization",
            solution=steps(
                f"Compute $Av = {latex(M)} {latex(v)} = {latex(Av)}$",
                f"We have $Av = {latex(Av)} = {eigenval} {latex(v)}$",
                f"Since $Av = \\lambda v$ with $\\lambda = {eigenval}$, $v$ is an eigenvector"
            ),
        )
    
    else:
        # Type 4: Identify which matrix is already diagonalized (1100-1200)
        d1, d2 = nonzero(-5, 5), nonzero(-5, 5)
        diag_matrix = Matrix([[d1, 0], [0, d2]])
        
        # Create a non-diagonal matrix
        a, b = nonzero(-4, 4), nonzero(-4, 4)
        non_diag = Matrix([[d1, b], [0, d2]])
        
        return problem(
            question=f"Which matrix is in diagonal form: $A = {latex(diag_matrix)}$ or $B = {latex(non_diag)}$? (Answer A or B)",
            answer="A",
            difficulty=(1100, 1200),
            topic="linear_algebra/diagonalization",
            solution=steps(
                "A matrix is in diagonal form if all off-diagonal entries are zero",
                f"Matrix $A = {latex(diag_matrix)}$ has only diagonal entries",
                f"Matrix $B = {latex(non_diag)}$ has off-diagonal entry ${b}$ in position (1,2)",
                "Therefore, $A$ is in diagonal form"
            ),
        )

emit(generate())

File: src/linear_algebra_diagonalization_easy.jl
# linear_algebra - diagonalization (easy)
# Generated: 2026-03-08T21:13:47.025796

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("linear_algebra/diagonalization")
    
    prob_type = choice([:eigenvalue_2x2, :eigenvector_2x2, :diagonal_verify, :diagonal_matrix_power])
    
    if prob_type == :eigenvalue_2x2
        # Generate 2x2 matrix with integer eigenvalues
        λ1 = randint(-8, 8)
        λ2 = randint(-8, 8)
        
        # Random invertible change of basis matrix
        p11, p12 = nonzero(-3, 3), nonzero(-3, 3)
        p21, p22 = nonzero(-3, 3), nonzero(-3, 3)
        
        # Ensure invertible (determinant nonzero)
        while p11*p22 - p12*p21 == 0
            p22 = nonzero(-3, 3)
        end
        
        # A = P D P^(-1), but we'll just construct directly
        a11 = λ1
        a22 = λ2
        a12 = randint(-5, 5)
        a21 = 0
        
        A = [[a11, a12], [a21, a22]]
        
        problem(
            question="Find the eigenvalues of the matrix \$A = $(fmt_matrix(A))\$. Enter your answer as a set.",
            answer=fmt_set(Set([λ1, λ2])),
            difficulty=(700, 900),
            answer_type="set",
            solution=steps(
                "Given matrix: \$A = $(fmt_matrix(A))\$",
                "For upper triangular matrix, eigenvalues are diagonal entries",
                sol("Eigenvalues", fmt_set(Set([λ1, λ2])))
            ),
            time=90
        )
        
    elseif prob_type == :eigenvector_2x2
        # Simple diagonal matrix with eigenvalue
        λ = nonzero(-7, 7)
        other_λ = nonzero(-7, 7)
        while other_λ == λ
            other_λ = nonzero(-7, 7)
        end
        
        A = [[λ, 0], [0, other_λ]]
        
        # Eigenvector for λ is [1, 0]
        v1 = 1
        v2 = 0
        
        problem(
            question="Find an eigenvector corresponding to eigenvalue \$\\lambda = $(λ)\$ for the matrix \$A = $(fmt_matrix(A))\$. Enter as a column vector.",
            answer=fmt_matrix([[v1], [v2]]),
            difficulty=(800, 1000),
            answer_type="matrix",
            solution=steps(
                "Given: \$A = $(fmt_matrix(A))\$, eigenvalue \$\\lambda = $(λ)\$",
                "Solve \$(A - \\lambda I)\\mathbf{v} = \\mathbf{0}\$",
                "\$A - $(λ)I = $(fmt_matrix([[0, 0], [0, other_λ - λ]]))\$",
                "First row gives \$0 = 0\$ (always true), second row gives \$$(other_λ - λ)v_2 = 0\$",
                "So \$v_2 = 0\$ and \$v_1\$ is free. Choose \$v_1 = 1\$",
                sol("Eigenvector", fmt_matrix([[v1], [v2]]))
            ),
            time=120
        )
        
    elseif prob_type == :diagonal_verify
        # Give a matrix and diagonal form, verify one eigenvalue
        λ1 = nonzero(-6, 6)
        λ2 = nonzero(-6, 6)
        
        # Upper triangular for simplicity
        a11 = λ1
        a22 = λ2
        a12 = nonzero(-4, 4)
        
        A = [[a11, a12], [0, a22]]
        
        problem(
            question="The matrix \$A = $(fmt_matrix(A))\$ has eigenvalues on its diagonal (it's upper triangular). What is the sum of all eigenvalues?",
            answer=λ1 + λ2,
            difficulty=(700, 900),
            solution=steps(
                "Given matrix: \$A = $(fmt_matrix(A))\$",
                "For triangular matrices, eigenvalues are the diagonal entries",
                "Eigenvalues: \$$(λ1), $(λ2)\$",
                sol("Sum", λ1 + λ2)
            ),
            time=60
        )
        
    else  # diagonal_matrix_power
        # Diagonal matrix raised to a power
        d1 = choice([-2, -1, 2, 3])
        d2 = choice([-2, -1, 2, 3])
        n = randint(2, 4)
        
        D = [[d1, 0], [0, d2]]
        
        Dn = [[d1^n, 0], [0, d2^n]]
        
        problem(
            question="If \$D = $(fmt_matrix(D))\$, compute \$D^{$(n)}\$.",
            answer=fmt_matrix(Dn),
            difficulty=(900, 1100),
            answer_type="matrix",
            solution=steps(
                "Given: \$D = $(fmt_matrix(D))\$",
                "For diagonal matrices, \$D^n\$ means raising each diagonal entry to the \$n\$-th power",
                "\$D^{$(n)} = $(fmt_matrix([[d1^n, 0], [0, d2^n]]))\$",
                sol("Answer", fmt_matrix(Dn))
            ),
            time=90
        )
    end
end